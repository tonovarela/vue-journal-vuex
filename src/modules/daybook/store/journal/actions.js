import journalApi from '@/api/journalApi'


export const loadEntries = async({commit})=>{
      const  {data}=  await journalApi.get('entries.json');
      if (!data){
        commit('setEntries', []);      
        return ;  
      }

      const entries = [];
      for(let id   of Object.keys(data)){
          entries.push( {id, ...data[id]});          
      }
      //console.log(entries);
      commit('setEntries', entries);      

}

export const updateEntry = async({commit},entry)=>{
    const { date, picture,text} = entry;
    const dataToSave = { date, picture,text};
    await journalApi.put(`entries/${entry.id}.json`, dataToSave);
    dataToSave.id =entry.id;
    commit('updateEntry',{...dataToSave});
}

export const createEntry = async({commit},entry)=>{
    const { date, picture,text} = entry;
    const dataToSave = { date, picture,text};
    const e =await journalApi.post(`entries.json`, dataToSave);
    const id =e.data.name;
    entry.id=id;
    commit('addEntry',{...entry});
}

export const deleteEntry = async({commit},id)=> {        
    await journalApi.delete(`entries/${id}.json`);
    commit('deleteEntry',id);
    
}