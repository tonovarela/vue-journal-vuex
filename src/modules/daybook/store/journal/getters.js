// export const myGetter = (state)=>{
    //return state

// }

export const getEntriesByTerm = ( /*state*/ )=>{
//    return state

}


export const getEntriesById = ( /*state*/ )=>{
    //return state

}