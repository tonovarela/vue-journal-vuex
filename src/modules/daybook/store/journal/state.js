export default ()=> ({
    isLoading:true,
    entries:[
      //   {
      //      //id:new Date().getTime(),
      //      date:new Date().toDateString(),
      //      id:'1',
      //       text:"Lorem, ipsum dolor sit amet consectetur adipisicing elit. Doloribus modi consequuntur ab blanditiis cumque ducimus illum totam at inventore laudantium sed, provident quisquam expedita a repellat praesentium ipsam velit optio!",
      //       picture:null
      //    },
      //    {
      //       //id:new Date().getTime()+1000,
      //       date:new Date().toDateString(),
      //       id:'2',
      //       text:"sds",
      //       picture:null
      //    },
      //    {
      //       //id:new Date().getTime()+2001,
      //       date:new Date().toDateString(),
      //       id:'3',
      //       text:"afff",
      //       picture:null
      //    },
      //    {
      //       //id:new Date().getTime()+3002,
      //       date:new Date().toDateString(),
      //       id:'4',
      //       text:"xx",
      //       picture:null
      //    }
    ]

})