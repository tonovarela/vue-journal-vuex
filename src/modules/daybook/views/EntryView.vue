<template>
<div v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
        <div>
            <span class="text-success fs-3 fw-bold">{{ day }}</span>
            <span class="mx-1 fs-3">{{ month }}</span>
            <span class="mx-2 fs-4 fw-light">
                {{ year }}
            </span>
        </div>

        <div> 

            <input accept="image/png, image/jpeg" v-show="false" ref="ImageSelector" type="file"  @change="onSelectedImage"   />
            <button  v-if="entry.id" @click="onDeleteEntry" class="btn btn-danger mx-2">
                Borrar
                <i class="fa fa-trash-alt"></i>
            </button>
            <button  @click="onSelectImage" class="btn btn-primary">
                Subir foto
                <i class="fa fa-upload"></i>
            </button>
        </div>
    </div>
    <hr />
    <div class="d-flex flex-column px-3 h-75">
        <textarea v-model="entry.text" class="form-control" placeholder="Â¿Que sucedio hoy?"></textarea>
    </div>
</div>

<img v-if="entry.picture && !localImage" :src="entry.picture" />

<img v-if="localImage" :src="localImage" class="img-thumbail" alt="entry-picture" />

<Fab icon="fa-save" @on:click="saveEntry"></Fab>
</template>

<script>
import {
    defineAsyncComponent
} from "vue";
import {
    mapGetters,
    mapActions
} from 'vuex';

import getDayMothYear from '../helpers/getDayMothYear'
import Swal from 'sweetalert2';
import uploadImage  from '../helpers/uploadImage';


export default {
    components: {
        Fab: defineAsyncComponent(() => import("../components/Fab")),
    },
    props: {
        id: {
            type: String,
            required: true
        },
    },

    data() {
        return {
            entry: null,
            localImage:null,
            file:null
        }
    },

    computed: {
        ...mapGetters('journal', ['getEntriesById']),

        day() {
            const {
                day
            } = getDayMothYear(this.entry.date)
            return day;
        },
        month() {
            const {
                month
            } = getDayMothYear(this.entry.date)
            return month;
        },
        year() {
            const {
                year
            } = getDayMothYear(this.entry.date)
            return year;
        }
    },
    methods: {
        ...mapActions('journal', ['updateEntry', 'createEntry','deleteEntry']),
        async saveEntry() {
            new Swal({
             title:"Espere por favor",
             allowOutsideClick:false
            });
            Swal.showLoading();
            this.entry.picture=await uploadImage(this.file);             
            if (this.id == "new") {
                await this.createEntry(this.entry);
                const id = this.entry.id;
                this.$router.push({
                    name: 'entry',
                    params: {
                        id
                    }
                })

            } else {
                this.updateEntry(this.entry);
            }
            Swal.fire('Guardado','Entrada registrada con exito','success');
            this.localImage= null;

        },

        async onDeleteEntry() {  
            
            const {isDenied} = await Swal.fire({
                title:"Esta seguro?",
                text:"Una vez borrado, no se puede recuperar",
                showDenyButton:true,
                confirmButtonText:"Si, estoy seguro"
            });
            if (isDenied){
                return;
            }
            new Swal({
                title:"Espere por favor",
                allowOutsideClick:false,
            });
            Swal.showLoading();
             await this.deleteEntry(this.entry.id)
             this.$router.push({
                 name:'no-entry'
             });  
             Swal.fire('Eliminado','Entrada eliminada con exito','success');           

        },
        onSelectImage(){
             this.$refs.ImageSelector.click();
        },  
        onSelectedImage(event){
            const [file]= event.target.files;
          if (!file){              
              return;
          }                
          this.file= file;
           const fr = new FileReader();
           fr.onload =() =>this.localImage =fr.result;           
          fr.readAsDataURL(file);
        },
        loadEntry() {
            let entry = {};
            if (this.id === "new") {
                entry = {
                    text: "",
                    date: new Date().getTime()
                }
            } else {
                entry = this.getEntriesById(this.id);
            }
            if (!entry) return this.$router.push({
                name: 'no-entry'
            });
            this.entry = entry;
        }
    },
    created() {
        this.loadEntry()
    },
    watch: {
        id() {
            this.loadEntry()
        }
    }
};
</script>

<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus {
        outline: none;
    }
}

img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>
