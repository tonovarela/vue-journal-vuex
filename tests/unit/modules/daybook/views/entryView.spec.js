

import { createStore } from 'vuex';
import { shallowMount } from '@vue/test-utils'
import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state';
import EntryView from '@/modules/daybook/views/EntryView.vue';
import Swal from 'sweetalert2';


jest.mock('sweetalert2',()=>({
    fire: jest.fn(),
    showLoading:jest.fn(),
    close: jest.fn()
}))
describe('Pruebas en el EntryView',()=>{
    const createVuexStore = (inicialState) => createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...inicialState }
            }
        }
    })
    const store = createVuexStore(journalState);
    let wrapper;
    let mockRouter;
    beforeEach(() => {    
         mockRouter = {
            push: jest.fn()
        }
         jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props:{
                id:'-Mz2XELbKHLiTt97ay1g'
                },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })   
    // test('Debe de sacar el usuario por que el id no existe',()=>{        
    //     const wrapper = shallowMount(EntryView, {        
    //         props:{
    //             id:'dsfdsf'
    //             },
    //         global: {        
    //             mocks: {
    //                 $router: mockRouter
    //             },
    //             plugins: [store]
    //         }
    //     })

    //     expect(mockRouter.push).toHaveBeenCalledWith({name:'no-entry'})

    // })
    test('Debe de mostrar la entrada correctamente',()=>{
        expect(wrapper.html()).toMatchSnapshot();
        expect(mockRouter.push).not.toHaveBeenCalledWith()
    })

    test('Debe de borrar la entrada y salir',(done)=>{
        Swal.fire.mockReturnValueOnce( Promise.resolve({isConfirmed:true}))
        wrapper.find('.btn-danger').trigger('click')

       expect(Swal.fire).toHaveBeenCalledWith({
        title: "Esta seguro?",
        text: "Una vez borrado, no se puede recuperar",
        showDenyButton: true,
        confirmButtonText: "Si, estoy seguro"
    })  
    // setTimeout(()=>{
    //     expect(mockRouter.push).toHaveBeenCalled()
    //     done()
    // },2000)

    

    })

})