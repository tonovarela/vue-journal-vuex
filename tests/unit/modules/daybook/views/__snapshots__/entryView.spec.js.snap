// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView Debe de mostrar la entrada correctamente 1`] = `
<div>
  <div class="entry-title d-flex justify-content-between p-2">
    <div><span class="text-success fs-3 fw-bold">25</span><span class="mx-1 fs-3">Marzo</span><span class="mx-2 fs-4 fw-light">2022 Viernes </span></div>
    <div><input accept="image/png, image/jpeg" type="file" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
  </div>
  <hr>
  <div class="d-flex flex-column px-3 h-75"><textarea class="form-control" placeholder="Â¿Que sucedio hoy?"></textarea></div>
</div><img src="https://res.cloudinary.com/duphx2ezk/image/upload/v1648253006/cb9rzvjlw2krjwg3rioc.png">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
